plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,482 @@
+fu_mbim_qdu_updater_mbim_device_open_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	if (!mbim_device_open_full_finish (MBIM_DEVICE (mbim_device), res, &ctx->error)) {
+		ctx->open_attempts--;
+		if (ctx->open_attempts == 0) {
+			g_clear_object (&ctx->mbim_device);
+			g_main_loop_quit (ctx->mainloop);
+			return;
+		}
+
+		/* retry */
+		g_clear_error (&ctx->error);
+		fu_mbim_qdu_updater_mbim_device_open_attempt (ctx);
+		return;
+	}
+
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static void
+fu_mbim_qdu_updater_mbim_device_open_attempt (OpenContext *ctx)
+{
+	MbimDeviceOpenFlags open_flags = MBIM_DEVICE_OPEN_FLAGS_NONE;
+
+	/* all communication through the proxy */
+	open_flags |= MBIM_DEVICE_OPEN_FLAGS_PROXY;
+
+	g_debug ("trying to open MBIM device...");
+	mbim_device_open_full (ctx->mbim_device, open_flags, 10, NULL,
+			       fu_mbim_qdu_updater_mbim_device_open_ready, ctx);
+}
+
+static void
+fu_mbim_qdu_updater_mbim_device_new_ready (GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	ctx->mbim_device = mbim_device_new_finish (res, &ctx->error);
+	if (ctx->mbim_device == NULL) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	fu_mbim_qdu_updater_mbim_device_open_attempt (ctx);
+}
+
+gboolean
+fu_mbim_qdu_updater_open (FuMbimQduUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	g_autoptr(GFile) mbim_device_file = g_file_new_for_path (self->mbim_port);
+	OpenContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = NULL,
+		.error = NULL,
+		.open_attempts = FU_MBIM_QDU_MAX_OPEN_ATTEMPTS,
+	};
+
+	mbim_device_new (mbim_device_file, NULL, fu_mbim_qdu_updater_mbim_device_new_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* either we have all device or otherwise error is set */
+	if (ctx.mbim_device != NULL) {
+		g_warn_if_fail (!ctx.error);
+		self->mbim_device = ctx.mbim_device;
+		/* success */
+		return TRUE;
+	}
+
+	g_warn_if_fail (ctx.error != NULL);
+	g_warn_if_fail (ctx.mbim_device == NULL);
+	g_propagate_error (error, ctx.error);
+	return FALSE;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	GError		*error;
+} CloseContext;
+
+static void
+fu_mbim_qdu_updater_mbim_device_close_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	CloseContext *ctx = (CloseContext *) user_data;
+
+	/* ignore errors when closing */
+	mbim_device_close_finish (MBIM_DEVICE (mbim_device), res, &ctx->error);
+	g_clear_object (&ctx->mbim_device);
+	g_main_loop_quit (ctx->mainloop);
+}
+
+gboolean
+fu_mbim_qdu_updater_close (FuMbimQduUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	CloseContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = g_steal_pointer (&self->mbim_device),
+		.error = NULL,
+	};
+
+	g_return_val_if_fail (ctx.mbim_device != NULL, TRUE);
```

Comment:
fu_mbim_qdu_updater_close() is also called when the open fails, so it may happen that `ctx.mbim_device` is NULL at this point, we shouldn't warn on this. Please change this to just:
```
if (ctx.mbim_device == NULL)
    return TRUE;
```

Reply:
OK, I will change it as you suggest, thanks.